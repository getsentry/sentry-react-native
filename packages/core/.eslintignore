# Ignore build generated dir
dist

# Ignore dangerfile
dangerfile.js

# Android Test App includes js in JVM test reports
RNSentryAndroidTester
