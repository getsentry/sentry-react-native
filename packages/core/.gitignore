/ts3.8
/README.md
/package.tgz
