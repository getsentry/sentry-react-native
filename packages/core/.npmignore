*

# Included to the final package
!LICENSE.md
!CHANGELOG.md
!README.md
!/dist/**/*
!ts3.8/**/*
!RNSentry.podspec
!sentry.gradle
!react-native.config.js
!/ios/**/*
!/android/**/*
!src/js/NativeRNSentry.ts
!scripts/collect-modules.sh
!scripts/copy-debugid.js
!scripts/has-sourcemap-debugid.js
!scripts/sentry-xcode.sh
!scripts/sentry-xcode-debug-files.sh
!scripts/sentry_utils.rb
!scripts/expo-upload-sourcemaps.js

# Metro
!/metro.js
!/metro.d.ts

# Expo
!/expo.js
!/expo.d.ts
!/app.plugin.js
!/plugin/build/**/*
