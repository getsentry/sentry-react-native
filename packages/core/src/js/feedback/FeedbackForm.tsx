import type { SendFeedbackParams } from '@sentry/core';
import { captureFeedback, getCurrentScope, lastEventId, logger } from '@sentry/core';
import * as React from 'react';
import type { KeyboardTypeOptions } from 'react-native';
import {
  Alert,
  Image,
  Keyboard,
  KeyboardAvoidingView,
  SafeAreaView,
  ScrollView,
  Text,
  TextInput,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View
} from 'react-native';

import { sentryLogo } from './branding';
import { defaultConfiguration } from './defaults';
import defaultStyles from './FeedbackForm.styles';
import type { FeedbackFormProps, FeedbackFormState, FeedbackFormStyles,FeedbackGeneralConfiguration, FeedbackTextConfiguration } from './FeedbackForm.types';
import { isValidEmail } from './utils';

/**
 * @beta
 * Implements a feedback form screen that sends feedback to Sentry using Sentry.captureFeedback.
 */
export class FeedbackForm extends React.Component<FeedbackFormProps, FeedbackFormState> {
  public static defaultProps: Partial<FeedbackFormProps> = {
    ...defaultConfiguration
  }

  public constructor(props: FeedbackFormProps) {
    super(props);

    const currentUser = {
      useSentryUser: {
        email: this.props?.useSentryUser?.email || getCurrentScope()?.getUser()?.email || '',
        name: this.props?.useSentryUser?.name || getCurrentScope()?.getUser()?.name || '',
      }
    }

    this.state = {
      isVisible: true,
      name: currentUser.useSentryUser.name,
      email: currentUser.useSentryUser.email,
      description: '',
    };
  }

  public handleFeedbackSubmit: () => void = () => {
    const { name, email, description } = this.state;
    const { onSubmitSuccess, onSubmitError, onFormSubmitted } = this.props;
    const text: FeedbackTextConfiguration = this.props;

    const trimmedName = name?.trim();
    const trimmedEmail = email?.trim();
    const trimmedDescription = description?.trim();

    if ((this.props.isNameRequired && !trimmedName) || (this.props.isEmailRequired && !trimmedEmail) || !trimmedDescription) {
      Alert.alert(text.errorTitle, text.formError);
      return;
    }

    if (this.props.shouldValidateEmail && (this.props.isEmailRequired || trimmedEmail.length > 0) && !isValidEmail(trimmedEmail)) {
      Alert.alert(text.errorTitle, text.emailError);
      return;
    }

    const attachments = this.state.filename && this.state.attachment
    ? [
        {
          filename: this.state.filename,
          data: this.state.attachment,
        },
      ]
    : undefined;

    const eventId = lastEventId();
    const userFeedback: SendFeedbackParams = {
      message: trimmedDescription,
      name: trimmedName,
      email: trimmedEmail,
      associatedEventId: eventId,
    };

    try {
      this.setState({ isVisible: false });
      captureFeedback(userFeedback, attachments ? { attachments } : undefined);
      onSubmitSuccess({ name: trimmedName, email: trimmedEmail, message: trimmedDescription, attachments: undefined });
      Alert.alert(text.successMessageText);
      onFormSubmitted();
    } catch (error) {
      const errorString = `Feedback form submission failed: ${error}`;
      onSubmitError(new Error(errorString));
      Alert.alert(text.errorTitle, text.genericError);
      logger.error(`Feedback form submission failed: ${error}`);
    }
  };

  public onScreenshotButtonPress: () => void = () => {
    if (!this.state.filename && !this.state.attachment) {
      const { onAddScreenshot } = { ...defaultConfiguration, ...this.props };
      onAddScreenshot((filename: string, attachement: Uint8Array) => {
        this.setState({ filename, attachment: attachement });
      });
    } else {
      this.setState({ filename: undefined, attachment: undefined });
    }
  }

  /**
   * Renders the feedback form screen.
   */
  public render(): React.ReactNode {
    const { name, email, description } = this.state;
    const { onFormClose } = this.props;
    const config: FeedbackGeneralConfiguration = this.props;
    const text: FeedbackTextConfiguration = this.props;
    const styles: FeedbackFormStyles = { ...defaultStyles, ...this.props.styles };
    const onCancel = (): void => {
      onFormClose();
      this.setState({ isVisible: false });
    }

    if (!this.state.isVisible) {
      return null;
    }

    return (
    <SafeAreaView style={[styles.container, { padding: 0 }]}>
      <KeyboardAvoidingView behavior={'padding'} style={[styles.container, { padding: 0 }]}>
        <ScrollView>
          <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
            <View style={styles.container}>
              <View style={styles.titleContainer}>
                <Text style={styles.title}>{text.formTitle}</Text>
                {config.showBranding && (
                  <Image
                    source={{ uri: sentryLogo }}
                    style={styles.sentryLogo}
                    testID='sentry-logo'
                  />
                )}
              </View>

              {config.showName && (
              <>
                <Text style={styles.label}>
                  {text.nameLabel}
                  {config.isNameRequired && ` ${text.isRequiredLabel}`}
                </Text>
                <TextInput
                  style={styles.input}
                  placeholder={text.namePlaceholder}
                  value={name}
                  onChangeText={(value) => this.setState({ name: value })}
                />
              </>
              )}

              {config.showEmail && (
              <>
                <Text style={styles.label}>
                  {text.emailLabel}
                  {config.isEmailRequired && ` ${text.isRequiredLabel}`}
                </Text>
                <TextInput
                  style={styles.input}
                  placeholder={text.emailPlaceholder}
                  keyboardType={'email-address' as KeyboardTypeOptions}
                  value={email}
                  onChangeText={(value) => this.setState({ email: value })}
                />
              </>
              )}

              <Text style={styles.label}>
                {text.messageLabel}
                {` ${text.isRequiredLabel}`}
              </Text>
              <TextInput
                style={[styles.input, styles.textArea]}
                placeholder={text.messagePlaceholder}
                value={description}
                onChangeText={(value) => this.setState({ description: value })}
                multiline
              />
              {config.enableScreenshot && (
                <TouchableOpacity style={styles.screenshotButton} onPress={this.onScreenshotButtonPress}>
                  <Text style={styles.screenshotText}>
                  {!this.state.filename && !this.state.attachment
                    ? text.addScreenshotButtonLabel
                    : text.removeScreenshotButtonLabel}
                  </Text>
                </TouchableOpacity>
              )}
              <TouchableOpacity style={styles.submitButton} onPress={this.handleFeedbackSubmit}>
                <Text style={styles.submitText}>{text.submitButtonLabel}</Text>
              </TouchableOpacity>

              <TouchableOpacity style={styles.cancelButton} onPress={onCancel}>
                <Text style={styles.cancelText}>{text.cancelButtonLabel}</Text>
              </TouchableOpacity>
            </View>
          </TouchableWithoutFeedback>
        </ScrollView>
      </KeyboardAvoidingView>
    </SafeAreaView>
    );
  }
}
