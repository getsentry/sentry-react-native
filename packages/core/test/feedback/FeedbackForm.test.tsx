import { captureFeedback } from '@sentry/core';
import { fireEvent, render, waitFor } from '@testing-library/react-native';
import * as React from 'react';
import { Alert } from 'react-native';

import { FeedbackForm } from '../../src/js/feedback/FeedbackForm';
import type { FeedbackFormProps, FeedbackFormStyles } from '../../src/js/feedback/FeedbackForm.types';

const mockOnFormClose = jest.fn();
const mockOnSubmitSuccess = jest.fn();
const mockOnFormSubmitted = jest.fn();
const mockOnSubmitError = jest.fn();
const mockGetUser = jest.fn(() => ({
  email: 'test@example.com',
  name: 'Test User',
}));

jest.spyOn(Alert, 'alert');

jest.mock('@sentry/core', () => ({
  ...jest.requireActual('@sentry/core'),
  captureFeedback: jest.fn(),
  getCurrentScope: jest.fn(() => ({
    getUser: mockGetUser,
  })),
  lastEventId: jest.fn(),
}));

const defaultProps: FeedbackFormProps = {
  onFormClose: mockOnFormClose,
  onSubmitSuccess: mockOnSubmitSuccess,
  onFormSubmitted: mockOnFormSubmitted,
  onSubmitError: mockOnSubmitError,
  formTitle: 'Feedback Form',
  nameLabel: 'Name Label',
  namePlaceholder: 'Name Placeholder',
  emailLabel: 'Email Label',
  emailPlaceholder: 'Email Placeholder',
  messageLabel: 'Message Label',
  messagePlaceholder: 'Message Placeholder',
  submitButtonLabel: 'Submit Button Label',
  cancelButtonLabel: 'Cancel Button Label',
  isRequiredLabel: '(is required label)',
  errorTitle: 'Error',
  formError: 'Please fill out all required fields.',
  emailError: 'The email address is not valid.',
  successMessageText: 'Feedback success',
  genericError: 'Generic error',
};

const customStyles: FeedbackFormStyles = {
  container: {
    backgroundColor: '#ffffff',
  },
  title: {
    fontSize: 20,
    color: '#ff0000',
  },
  label: {
    fontSize: 15,
    color: '#00ff00',
  },
  input: {
    height: 50,
    borderColor: '#0000ff',
    fontSize: 13,
    color: '#000000',
  },
  textArea: {
    height: 50,
    color: '#00ff00',
  },
  submitButton: {
    backgroundColor: '#ffff00',
  },
  submitText: {
    color: '#ff0000',
    fontSize: 12,
  },
  cancelButton: {
    paddingVertical: 10,
  },
  cancelText: {
    color: '#ff0000',
    fontSize: 10,
  },
};

describe('FeedbackForm', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('matches the snapshot with default configuration', () => {
    const { toJSON } = render(<FeedbackForm/>);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches the snapshot with custom texts', () => {
    const { toJSON } = render(<FeedbackForm {...defaultProps}/>);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches the snapshot with custom styles', () => {
    const customStyleProps = {styles: customStyles};
    const { toJSON } = render(<FeedbackForm {...customStyleProps}/>);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly', () => {
    const { getByPlaceholderText, getByText, getByTestId } = render(<FeedbackForm {...defaultProps} />);

    expect(getByText(defaultProps.formTitle)).toBeTruthy();
    expect(getByTestId('sentry-logo')).toBeTruthy(); // default showBranding is true
    expect(getByText(defaultProps.nameLabel)).toBeTruthy();
    expect(getByPlaceholderText(defaultProps.namePlaceholder)).toBeTruthy();
    expect(getByText(defaultProps.emailLabel)).toBeTruthy();
    expect(getByPlaceholderText(defaultProps.emailPlaceholder)).toBeTruthy();
    expect(getByText(`${defaultProps.messageLabel } ${  defaultProps.isRequiredLabel}`)).toBeTruthy();
    expect(getByPlaceholderText(defaultProps.messagePlaceholder)).toBeTruthy();
    expect(getByText(defaultProps.submitButtonLabel)).toBeTruthy();
    expect(getByText(defaultProps.cancelButtonLabel)).toBeTruthy();
  });

  it('does not render the sentry logo when showBranding is false', () => {
    const { queryByTestId } = render(<FeedbackForm {...defaultProps} showBranding={false} />);

    expect(queryByTestId('sentry-logo')).toBeNull();
  });

  it('name and email are prefilled when sentry user is set', () => {
    const { getByPlaceholderText } = render(<FeedbackForm {...defaultProps} />);

    const nameInput = getByPlaceholderText(defaultProps.namePlaceholder);
    const emailInput = getByPlaceholderText(defaultProps.emailPlaceholder);

    expect(nameInput.props.value).toBe('Test User');
    expect(emailInput.props.value).toBe('test@example.com');
  });

  it('ensure getUser is called only after the component is rendered', () => {
    // Ensure getUser is not called before render
    expect(mockGetUser).not.toHaveBeenCalled();

    // Render the component
    render(<FeedbackForm />);

    // After rendering, check that getUser was called twice (email and name)
    expect(mockGetUser).toHaveBeenCalledTimes(2);
  });

  it('shows an error message if required fields are empty', async () => {
    const { getByText } = render(<FeedbackForm {...defaultProps} />);

    fireEvent.press(getByText(defaultProps.submitButtonLabel));

    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledWith(defaultProps.errorTitle, defaultProps.formError);
    });
  });

  it('shows an error message if the email is not valid and the email is required', async () => {
    const withEmailProps = {...defaultProps, ...{isEmailRequired: true}};
    const { getByPlaceholderText, getByText } = render(<FeedbackForm {...withEmailProps} />);

    fireEvent.changeText(getByPlaceholderText(defaultProps.namePlaceholder), 'John Doe');
    fireEvent.changeText(getByPlaceholderText(defaultProps.emailPlaceholder), 'not-an-email');
    fireEvent.changeText(getByPlaceholderText(defaultProps.messagePlaceholder), 'This is a feedback message.');

    fireEvent.press(getByText(defaultProps.submitButtonLabel));

    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledWith(defaultProps.errorTitle, defaultProps.emailError);
    });
  });

  it('calls captureFeedback when the form is submitted successfully', async () => {
    const { getByPlaceholderText, getByText } = render(<FeedbackForm {...defaultProps} />);

    fireEvent.changeText(getByPlaceholderText(defaultProps.namePlaceholder), 'John Doe');
    fireEvent.changeText(getByPlaceholderText(defaultProps.emailPlaceholder), 'john.doe@example.com');
    fireEvent.changeText(getByPlaceholderText(defaultProps.messagePlaceholder), 'This is a feedback message.');

    fireEvent.press(getByText(defaultProps.submitButtonLabel));

    await waitFor(() => {
      expect(captureFeedback).toHaveBeenCalledWith({
        message: 'This is a feedback message.',
        name: 'John Doe',
        email: 'john.doe@example.com',
      });
    });
  });

  it('shows success message when the form is submitted successfully', async () => {
    const { getByPlaceholderText, getByText } = render(<FeedbackForm {...defaultProps} />);

    fireEvent.changeText(getByPlaceholderText(defaultProps.namePlaceholder), 'John Doe');
    fireEvent.changeText(getByPlaceholderText(defaultProps.emailPlaceholder), 'john.doe@example.com');
    fireEvent.changeText(getByPlaceholderText(defaultProps.messagePlaceholder), 'This is a feedback message.');

    fireEvent.press(getByText(defaultProps.submitButtonLabel));

    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledWith(defaultProps.successMessageText);
    });
  });

  it('shows an error message when there is a an error in captureFeedback', async () => {
    (captureFeedback as jest.Mock).mockImplementationOnce(() => {
      throw new Error('Test error');
    });

    const { getByPlaceholderText, getByText } = render(<FeedbackForm {...defaultProps} />);

    fireEvent.changeText(getByPlaceholderText(defaultProps.namePlaceholder), 'John Doe');
    fireEvent.changeText(getByPlaceholderText(defaultProps.emailPlaceholder), 'john.doe@example.com');
    fireEvent.changeText(getByPlaceholderText(defaultProps.messagePlaceholder), 'This is a feedback message.');

    fireEvent.press(getByText(defaultProps.submitButtonLabel));

    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledWith(defaultProps.errorTitle, defaultProps.genericError);
    });
  });

  it('calls onSubmitError when there is an error', async () => {
    (captureFeedback as jest.Mock).mockImplementationOnce(() => {
      throw new Error('Test error');
    });

    const { getByPlaceholderText, getByText } = render(<FeedbackForm {...defaultProps} />);

    fireEvent.changeText(getByPlaceholderText(defaultProps.namePlaceholder), 'John Doe');
    fireEvent.changeText(getByPlaceholderText(defaultProps.emailPlaceholder), 'john.doe@example.com');
    fireEvent.changeText(getByPlaceholderText(defaultProps.messagePlaceholder), 'This is a feedback message.');

    fireEvent.press(getByText(defaultProps.submitButtonLabel));

    await waitFor(() => {
      expect(mockOnSubmitError).toHaveBeenCalled();
    });
  });

  it('calls onSubmitSuccess when the form is submitted successfully', async () => {
    const { getByPlaceholderText, getByText } = render(<FeedbackForm {...defaultProps} />);

    fireEvent.changeText(getByPlaceholderText(defaultProps.namePlaceholder), 'John Doe');
    fireEvent.changeText(getByPlaceholderText(defaultProps.emailPlaceholder), 'john.doe@example.com');
    fireEvent.changeText(getByPlaceholderText(defaultProps.messagePlaceholder), 'This is a feedback message.');

    fireEvent.press(getByText(defaultProps.submitButtonLabel));

    await waitFor(() => {
      expect(mockOnSubmitSuccess).toHaveBeenCalled();
    });
  });

  it('calls onFormSubmitted when the form is submitted successfully', async () => {
    const { getByPlaceholderText, getByText } = render(<FeedbackForm {...defaultProps} />);

    fireEvent.changeText(getByPlaceholderText(defaultProps.namePlaceholder), 'John Doe');
    fireEvent.changeText(getByPlaceholderText(defaultProps.emailPlaceholder), 'john.doe@example.com');
    fireEvent.changeText(getByPlaceholderText(defaultProps.messagePlaceholder), 'This is a feedback message.');

    fireEvent.press(getByText(defaultProps.submitButtonLabel));

    await waitFor(() => {
      expect(mockOnFormSubmitted).toHaveBeenCalled();
    });
  });

  it('calls onFormClose when the cancel button is pressed', () => {
    const { getByText } = render(<FeedbackForm {...defaultProps} />);

    fireEvent.press(getByText(defaultProps.cancelButtonLabel));

    expect(mockOnFormClose).toHaveBeenCalled();
  });
});
