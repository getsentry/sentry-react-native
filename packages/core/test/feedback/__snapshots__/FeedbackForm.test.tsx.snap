// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackForm matches the snapshot with custom styles 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "backgroundColor": "#ffffff",
      },
      {
        "padding": 0,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        [
          {
            "backgroundColor": "#ffffff",
          },
          {
            "padding": 0,
          },
        ],
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView>
      <View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#ffffff",
            }
          }
        >
          <Text
            style={
              {
                "color": "#ff0000",
                "fontSize": 20,
              }
            }
          >
            Report a Bug
          </Text>
          <Text
            style={
              {
                "color": "#00ff00",
                "fontSize": 15,
              }
            }
          >
            Name
          </Text>
          <TextInput
            onChangeText={[Function]}
            placeholder="Your Name"
            style={
              {
                "borderColor": "#0000ff",
                "color": "#000000",
                "fontSize": 13,
                "height": 50,
              }
            }
            value="Test User"
          />
          <Text
            style={
              {
                "color": "#00ff00",
                "fontSize": 15,
              }
            }
          >
            Email
          </Text>
          <TextInput
            keyboardType="email-address"
            onChangeText={[Function]}
            placeholder="your.email@example.org"
            style={
              {
                "borderColor": "#0000ff",
                "color": "#000000",
                "fontSize": 13,
                "height": 50,
              }
            }
            value="test@example.com"
          />
          <Text
            style={
              {
                "color": "#00ff00",
                "fontSize": 15,
              }
            }
          >
            Description
             (required)
          </Text>
          <TextInput
            multiline={true}
            onChangeText={[Function]}
            placeholder="What's the bug? What did you expect?"
            style={
              [
                {
                  "borderColor": "#0000ff",
                  "color": "#000000",
                  "fontSize": 13,
                  "height": 50,
                },
                {
                  "color": "#00ff00",
                  "height": 50,
                },
              ]
            }
            value=""
          />
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#ffff00",
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#ff0000",
                  "fontSize": 12,
                }
              }
            >
              Send Bug Report
            </Text>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
                "paddingVertical": 10,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#ff0000",
                  "fontSize": 10,
                }
              }
            >
              Cancel
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;

exports[`FeedbackForm matches the snapshot with custom texts 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "backgroundColor": "#ffffff",
        "flex": 1,
        "padding": 20,
      },
      {
        "padding": 0,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        [
          {
            "backgroundColor": "#ffffff",
            "flex": 1,
            "padding": 20,
          },
          {
            "padding": 0,
          },
        ],
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView>
      <View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#ffffff",
              "flex": 1,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              {
                "color": "#2b2233",
                "fontSize": 24,
                "fontWeight": "bold",
                "marginBottom": 20,
                "textAlign": "center",
              }
            }
          >
            Feedback Form
          </Text>
          <Text
            style={
              {
                "color": "#2b2233",
                "fontSize": 16,
                "marginBottom": 4,
              }
            }
          >
            Name Label
          </Text>
          <TextInput
            onChangeText={[Function]}
            placeholder="Name Placeholder"
            style={
              {
                "borderColor": "rgba(41, 35, 47, 0.13)",
                "borderRadius": 5,
                "borderWidth": 1,
                "color": "#2b2233",
                "fontSize": 16,
                "height": 50,
                "marginBottom": 15,
                "paddingHorizontal": 10,
              }
            }
            value="Test User"
          />
          <Text
            style={
              {
                "color": "#2b2233",
                "fontSize": 16,
                "marginBottom": 4,
              }
            }
          >
            Email Label
          </Text>
          <TextInput
            keyboardType="email-address"
            onChangeText={[Function]}
            placeholder="Email Placeholder"
            style={
              {
                "borderColor": "rgba(41, 35, 47, 0.13)",
                "borderRadius": 5,
                "borderWidth": 1,
                "color": "#2b2233",
                "fontSize": 16,
                "height": 50,
                "marginBottom": 15,
                "paddingHorizontal": 10,
              }
            }
            value="test@example.com"
          />
          <Text
            style={
              {
                "color": "#2b2233",
                "fontSize": 16,
                "marginBottom": 4,
              }
            }
          >
            Message Label
             (is required label)
          </Text>
          <TextInput
            multiline={true}
            onChangeText={[Function]}
            placeholder="Message Placeholder"
            style={
              [
                {
                  "borderColor": "rgba(41, 35, 47, 0.13)",
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "color": "#2b2233",
                  "fontSize": 16,
                  "height": 50,
                  "marginBottom": 15,
                  "paddingHorizontal": 10,
                },
                {
                  "color": "#2b2233",
                  "height": 100,
                  "textAlignVertical": "top",
                },
              ]
            }
            value=""
          />
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "rgba(88, 74, 192, 1)",
                "borderRadius": 5,
                "marginBottom": 10,
                "opacity": 1,
                "paddingVertical": 15,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#ffffff",
                  "fontSize": 18,
                }
              }
            >
              Submit Button Label
            </Text>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "opacity": 1,
                "paddingVertical": 15,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#2b2233",
                  "fontSize": 16,
                }
              }
            >
              Cancel Button Label
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;

exports[`FeedbackForm matches the snapshot with default configuration 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "backgroundColor": "#ffffff",
        "flex": 1,
        "padding": 20,
      },
      {
        "padding": 0,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        [
          {
            "backgroundColor": "#ffffff",
            "flex": 1,
            "padding": 20,
          },
          {
            "padding": 0,
          },
        ],
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView>
      <View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#ffffff",
              "flex": 1,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              {
                "color": "#2b2233",
                "fontSize": 24,
                "fontWeight": "bold",
                "marginBottom": 20,
                "textAlign": "center",
              }
            }
          >
            Report a Bug
          </Text>
          <Text
            style={
              {
                "color": "#2b2233",
                "fontSize": 16,
                "marginBottom": 4,
              }
            }
          >
            Name
          </Text>
          <TextInput
            onChangeText={[Function]}
            placeholder="Your Name"
            style={
              {
                "borderColor": "rgba(41, 35, 47, 0.13)",
                "borderRadius": 5,
                "borderWidth": 1,
                "color": "#2b2233",
                "fontSize": 16,
                "height": 50,
                "marginBottom": 15,
                "paddingHorizontal": 10,
              }
            }
            value="Test User"
          />
          <Text
            style={
              {
                "color": "#2b2233",
                "fontSize": 16,
                "marginBottom": 4,
              }
            }
          >
            Email
          </Text>
          <TextInput
            keyboardType="email-address"
            onChangeText={[Function]}
            placeholder="your.email@example.org"
            style={
              {
                "borderColor": "rgba(41, 35, 47, 0.13)",
                "borderRadius": 5,
                "borderWidth": 1,
                "color": "#2b2233",
                "fontSize": 16,
                "height": 50,
                "marginBottom": 15,
                "paddingHorizontal": 10,
              }
            }
            value="test@example.com"
          />
          <Text
            style={
              {
                "color": "#2b2233",
                "fontSize": 16,
                "marginBottom": 4,
              }
            }
          >
            Description
             (required)
          </Text>
          <TextInput
            multiline={true}
            onChangeText={[Function]}
            placeholder="What's the bug? What did you expect?"
            style={
              [
                {
                  "borderColor": "rgba(41, 35, 47, 0.13)",
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "color": "#2b2233",
                  "fontSize": 16,
                  "height": 50,
                  "marginBottom": 15,
                  "paddingHorizontal": 10,
                },
                {
                  "color": "#2b2233",
                  "height": 100,
                  "textAlignVertical": "top",
                },
              ]
            }
            value=""
          />
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "rgba(88, 74, 192, 1)",
                "borderRadius": 5,
                "marginBottom": 10,
                "opacity": 1,
                "paddingVertical": 15,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#ffffff",
                  "fontSize": 18,
                }
              }
            >
              Send Bug Report
            </Text>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "opacity": 1,
                "paddingVertical": 15,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#2b2233",
                  "fontSize": 16,
                }
              }
            >
              Cancel
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
